README
Net-Google-SafeBrowsing4
========================

Net::Google::SafeBrowsing4 implements the Google Safe Browsing v4 Local API

The library passes all the unit tests listed in the API documentation.

The Google Safe Browsing database must be stored and managed locally.
Net::Google::SafeBrowsing4::Storage::File uses files as the storage back-end.
Other storage mechanisms (databases, memory, etc.) can be added and used transparently with this module.


The source code is available on github at https://github.com/juliensobrier/Net-Google-SafeBrowsing4.


INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  constant
  Carp
  Digest::SHA
  Exporter
  HTTP::Message
  JSON::XS
  List::Util
  LWP::UserAgent
  MIME::Base64
  Net::IP::Lite
  Path::Tiny
  Storable
  Test::LWP::UserAgent
  Test::More
  Text::Trim
  Time::HiRes
  URI


COPYRIGHT AND LICENCE

Copyright (C) 2017 by Julien Sobrier

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension Net::Google::SafeBrowsing4.

0.7 Sun November 26 21:21:00 2014
 - Save lists locally - issue 16

0.6 Fri August 18 19:10:00 2017
	- Fix manifest file

0.5  Wed August 16 20:15:00 2017
	- use binary search to get local prefix (faster) - issue #8
	- update last_error if full hash request failed - issue #7

0.4  Wed August 176 19:10:00 2017
	- Fix manifest file

0.3  Sat March 05 17:38:00 2017
	- Lookup multiple URLs at once

0.2  Tue November 29 21:12:37 2016
	- URI normalization overhaul
	
0.1  Sat October 08 14:52:44 2016
	- initial version


Makefile.PL
#use 5.010001;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
	NAME			=> 'Net::Google::SafeBrowsing4',
	VERSION_FROM	=> 'lib/Net/Google/SafeBrowsing4.pm',
	PREREQ_PM		=> {
		constant		=> 0,
		Carp			=> 0,
		Digest::SHA		=> 0,
		Exporter		=> 0,
		HTTP::Message		=> 0,
		JSON::XS		=> 0,
		List::Util		=> 0,
		LWP::UserAgent		=> 0,
		MIME::Base64		=> 0,
		Net::IP::Lite		=> 0,
		Path::Tiny		=> 0,
		Storable		=> 0,
		Test::More		=> 0.92,
		Test::Pod::Coverage	=> 0,
		Test::LWP::UserAgent	=> 0,
		Text::Trim		=> 0,
		Time::HiRes		=> 0,
		URI			=> 0,
	},
	META_MERGE => {
		'meta-spec' => { version => 2 },
			resources => {
				repository => {
				type => 'git',
				url => 'https://github.com/juliensobrier/Net-Google-SafeBrowsing4.git',
				web => 'https://github.com/juliensobrier/Net-Google-SafeBrowsing4',
			},
		},
	},
	(
		$] >= 5.005 ?     ## Add these new keywords supported since 5.005
			(
				ABSTRACT_FROM	=> 'lib/Net/Google/SafeBrowsing4.pm', # retrieve abstract from module
				AUTHOR			=> 'Julien Sobrier <julien@sobrier.net>'
			) : ()
	),
);


lib/Net/Google/SafeBrowsing4/URI.pm
package Net::Google::SafeBrowsing4::URI;

use strict;
use warnings;

use Digest::SHA qw(sha256);
use Net::IP::Lite qw();
use URI;

=encoding utf-8

=head1 NAME

Net::Google::SafeBrowsing4::URI - URI management Class for the Google SafeBrowsing (version 4) API.


=head1 SYNOPSIS

	use Net::Google::SafeBrowsing4::URI;

	my $gsb_uri = Net::Google::SafeBrowsing4::URI->new('http://my.example.site:80/path/to/file.html?query=param#fragment');
	my @uris = $gsb_uri->generate_lookupuris();

=head1 DESCRIPTION

Net::Google::SafeBrowsing4::URI takes care of normalizing URLs, extracting suffix/prefix expressions, calculating hashes.

=head1 METHODS

=over

=item new

=back

	my $gsb_uri = Net::Google::SafeBrowsing4::URI->new('http://my.example.site:80/path/to/file.html?query=param#fragment');

=over

Initializes the object.

Arguments:

=over

=item $uri The URL to parse

=back

=back

=cut

sub new {
	my $class = shift;
	my @args = @_;

	if ((scalar(@args) == 0) || !$args[0]) {
		return undef;
	}

	my $self = {
		rawuri => $args[0],
	};

	bless($self, $class);
	return $self->_normalize() ? $self : undef;
}

=over

=item as_string

Returns the normalized URI as string.

=back

=cut

sub as_string {
	my $self = shift;

	return '' . $self->{uri};
}

=item generate_lookupuris

Generates all partial/full URIs supported by Google SafeBrowsing. See "suffix/prefix expressions" topic in GSBv4 API reference.
Returns a list of L<Net::Google::SafeBrowsing4::URI> objects.

=cut

sub generate_lookupuris {
	my $self = shift;
	my @uris = ();

	$self->as_string() =~ /^(https?:\/\/)([^\/]+)(\/[^\?]*)(\??.*)$/i;
	my ($scheme, $host, $path, $query) = ($1, $2, $3, $4);

	# Collect host suffixes
	my @domains = ();
	if ($host !~ /^\d+\.\d+\.\d+\.\d+$/) {
		my @parts = split(/\./, $host);
		splice(@parts, 0, -6); # take 5 top most components

		while (scalar(@parts) > 2) {
			shift(@parts);
			push(@domains, join(".", @parts));
		}
	}
	push(@domains, $host);

	# Collect path & query prefixes
	my @paths = ();
	if ($path ne '/') {
		my @parts = split(/\//, $path);
		my $part_count = scalar(@parts);
		$part_count = $part_count > 4 ? 4 : $part_count - 1; # limit to 4
		my $previous = "";
		# Skip the first entry in @parts as it is always an empty string
		for (my $i = 1; $i < $part_count; $i++) {
			$previous .= "/" . $parts[$i] ."/";
			push(@paths, $previous);
		}
	}
	push(@paths, $path);
	if ($query =~ /^\?./) {
		push(@paths, $path . $query);
	}

	# Assemble the list of Net::Google::SafeBrowsing4::URI objects
	foreach my $domain (@domains) {
		foreach my $path (@paths) {
			my $gsb_uri = Net::Google::SafeBrowsing4::URI->new($scheme . $domain . $path);
			# @TODO Sub-URI of a valid URI should be a valid URI. Condition should not be necessary.
			if (defined($gsb_uri)) {
				push(@uris, $gsb_uri);
			}
		}
	}

	return @uris;
}

=item hash

Generates the SHA-256 hash of the URI (with scheme removed).

=cut

sub hash {
	my $self = shift;

	my $uri = $self->as_string();
	$uri =~ s/^https?:\/\///i;

	return sha256($uri);
}

=head1 PRIVATE METHODS

=over

=item _normalize

Parses and normalizes the URI.

=back

=cut

sub _normalize {
	my $self = shift;
	my $modified_rawuri = $self->{rawuri};

	# Remove third and more slashes after the scheme
	$modified_rawuri =~ s/^(\s*https?:\/\/)\/+/$1/si;
	# Remove any Tab, CR, LF characters from the URI
	$modified_rawuri =~ s/[\r\n\t]+//sgi;
	# Recursive percent-unescape (everything but '#' not to confuse URI parser)
	while ($modified_rawuri =~ s{%(?!23)([[:xdigit:]]{2})}{chr(hex($1))}esg) { }

	# Parse URI
	my $uri_obj = URI->new($modified_rawuri);
	if (ref($uri_obj) !~ /^URI::https?$/) {
		if (!$uri_obj->scheme() || (!$uri_obj->has_recognized_scheme() && $modified_rawuri =~ /^[^:]+:\d{1,5}(?:\/|$)/)) {
			$uri_obj = URI->new("http://" . $modified_rawuri);
		}
	}
	# Only http and https URIs are supported
	if (ref($uri_obj) !~ /^URI::https?$/) {
		return undef;
	}

	# Remove userinfo
	$uri_obj->userinfo(undef);
	# Remove port
	$uri_obj->port(undef);
	# Remove Fragment
	$uri_obj->fragment(undef);

	# Host modifications
	my $modified_host = $uri_obj->host();
	# Host part cannot be empty
	if ($modified_host =~ /^\s*$/) {
		return undef;
	}
	# Collapse consecutive dots into one
	$modified_host =~ s/\.\.+/\./sg;
	# Remove leading and trailing dot
	$modified_host =~ s/^\.|\.$//sg;

	# IPv4 canonicalizations
	$modified_host = _normalize_ip($modified_host);
	if (!defined($modified_host)) {
		return undef;
	}
	$uri_obj->host($modified_host);

	# Empty/separator-only host
	if ($uri_obj->host() =~ /^[\s.\/]*$/) {
		return undef;
	}

	# Numeric TLD (and not IPv4)
	if ($uri_obj->host() =~ /\.\d[^\.]*$/
	 && $uri_obj->host() !~ /^(?:2(?:5[0-5]|[0-4][0-9])|[01]?[0-9]{1,2})(?:\.(?:2(?:5[0-5]|[0-4][0-9])|[01]?[0-9]{1,2})){3}$/) {
		return undef;
	}

	my $modified_path = $uri_obj->path();
	# Eliminate current directory /./ parts
	while ($modified_path =~ s/\/\.(?:\/|$)/\//sg) {};
	# Eliminate parent directory /something/./ parts
	while ($modified_path =~ s/\/[^\/]+\/\.\.(?:\/|$)/\//sg) {};
	# Eliminate double // slashes from path
	$modified_path =~ s/\/\/+/\//sg;
	if ($modified_path eq '') {
		$modified_path = '/';
	}
	$uri_obj->path($modified_path);

	# Fix some percent encoding
	my $modified_path_query = $uri_obj->path_query();
	# Fix lone percent signs %
	$modified_path_query =~ s/%(?![[:xdigit:]]{2})/%25/sg;
	$uri_obj->path_query($modified_path_query);

	my $canonical = $uri_obj->canonical();
	# Fix caret escaping
	$canonical=~ s/%5E/\^/sg;

	$self->{uri} = $canonical;

	return $self->{uri};
}

=head1 PRIVATE FUNCTIONS

=over

=item _normalize_ip

Function for recognising various IPv4 formatted addresses and convert them to I<dotted-decimal-quad> format (111.11.1.1)

Arguments:

=over

=item $host

Hostname to parse as IP Address

=back

=back

=cut

sub _normalize_ip {
	my $host = shift;

	# Shortcut: If it doesn't look like an IPv4, then return early
	if ($host !~ /^(?:0x[[:xdigit:]]+|\d+)(?:\.(?:0x[[:xdigit:]]+|\d+))*$/si) {
		return $host;
	}

	# Most formats are detected and converted by Net::IP::Lite
	my $ip = Net::IP::Lite->new($host);
	if ($ip) {
		return $ip->transform();
	}

	# One and two dots case is missing: xxx.xxxxxxxxxx, xxx.xxx.xxxxxx
	my $bits = 32;
	my @segments = split(/\./, $host);
	my $segment_count = scalar(@segments);

	my $decimal = 0;
	for (my $i = 0; $i < $segment_count; $i++) {
		my $is_last_segment = $i >= $segment_count - 1;
		my $segment = _parse_ipv4_segment($segments[$i], !$is_last_segment ? 8 : $bits);
		if (!defined($segment)) {
			return undef;
		}
		$bits -= 8;
		$decimal +=  $segment << (!$is_last_segment ? $bits : 0);
	}

	$ip = Net::IP::Lite->new($decimal);
	return $ip->transform();
}

=over

=item _parse_ipv4_segment

=back

	my $decimal = _parse_ipv4_part($segment, $bits)

=over

Transforms one IPv4 segment to decimal with range checking.

Arguments:

=over

=item $segment

Decimal/octal/hexadecimal value to parse

=item $bits

Bit length for range checking

=back

=back

=cut

sub _parse_ipv4_segment {
	my $segment = shift;
	my $bits = shift;
	my $decimal;

	if ($segment =~ /^0+([0-7]{0,10}|[0-3][0-7]{10})$/) {
		$decimal = oct($1);
	}
	elsif ($segment =~ /^0x0*([[:xdigit:]]{1,8})$/si) {
		$decimal = hex($1);
	}
	elsif ($segment =~ /^[1-9]\d+$/) {
		$decimal = $segment;
	}
	else {
		return undef;
	}

	if ($decimal >= (1 << $bits)) {
		return undef;
	}
	return $decimal;
}

=head1 BUGS

Some URI normalizatuion cases are still missing:

=over

=item Highbit characters in hostname are punycoded, not percent encoded.

Google's GO Client percent-escapes based on the hostname is a valid unicode string or not

=back

=head1 AUTHORS

Tamás Fehérvári, E<lt>geever@users.sourceforge.net<gt>
Julien Sobrier, E<lt>julien@sobrier.netE<gt>,

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2016 by Julien Sobrier, Tamás Fehérvári

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

=cut

1;


lib/Net/Google/SafeBrowsing4/Storage.pm
package Net::Google::SafeBrowsing4::Storage;

# ABSTRACT: Base class for storing the Google Safe Browsing v4 database

use strict;
use warnings;

our $VERSION = '0.1';

=head1 NAME

Net::Google::SafeBrowsing4::Storage - Base class for storing the Google Safe Browsing v4 database

=head1 SYNOPSIS

	package Net::Google::SafeBrowsing4::Storage::File;

	use base qw(Net::Google::SafeBrowsing4::Storage);

=head1 DESCRIPTION

This is the base class for implementing a storage mechanism for the Google Safe Browsing v4 database. See L<Net::Google::SafeBrowsing4::Storage::File> for an example of implementation.

This module cannot be used on its own as it does not actually store anything. All public methods should redefined.

=cut


=head1 CONSTRUCTOR

=head2 new()

	Create a Net::Google::SafeBrowsing4::Storage object

	my $storage => Net::Google::SafeBrowsing4::Storage->new(
		# Constructor parameters vary based on the implementation
		...
	);

=cut

sub new {
	...
}

=head1 PUBLIC FUNCTIONS

=over 4

=back

=head2 save()

Add chunk information to the local database

  $storage->save(add => [...], remove => [...], state => '...', list => { threatType => ..., threatEntryType => ..., platformType => ... });

Return the new list of local hashes.


Arguments

=over 4

=item override

Optional. override the local list of hashes. 0 by default (do not override)

=item add

Optional. List of hashes to add.

=item remove

Optional. List of hash indexes to remove.

=item state

Optional. New list state.

=item list

Required. Google Safe Browsing list.

=back

=cut

sub save {
	...
}


=head2 reset()

Remove all local data.

	$storage->reset(list => { threatType => ..., threatEntryType => ..., platformType => ... });


Arguments

=over 4

=item list

Required. Google Safe Browsing list.

=back

No return value

=cut

sub reset {
	...
}


=head2 next_update()

Ge the timestamp when the local database update is allowed.

	my $next = $storage->next_update();


No arguments

=cut

sub next_update {
	...
}


=head2 get_state()

Return the current state of the list.

	my $state = $storage->get_state(list => { threatType => ..., threatEntryType => ..., platformType => ... });


Arguments

=over 4

=item list

Required. Google Safe Browsing list.

=back


=cut

sub get_state {
	...
}



=head2 get_prefixes()

Return the list of prefxies that match a full hash for a given list.

	my @prefixes = $storage->get_prefixes(hashes => [...], list => { threatType => ..., threatEntryType => ..., platformType => ... });


Arguments

=over 4

=item list

Required. Google Safe Browsing list.

=back

=item hashes

Required. List of full hashes.

=back


=cut

sub get_prefixes {
	...
}

=head2 updated()

Save information about a successful database update

	$storage->updated('time' => time(), next => time() + 1800);


Arguments

=over 4

=item time

Required. Time of the update.

=item next

Required. Time of the next update allowed.

=back


No return value

=cut

sub updated {
	...
}


=head2 get_full_hashes()

Return a list of full hashes

	$storage->get_full_hashes(hash => AAAAAAAA..., lists => [{ threatType => '...', threatEntryType => '...', platformType => '...' }]);


Arguments

=over 4

=item hash

Required. 32-bit hash


=item lists

Required. Google Safe Browsing lists

=back

Return value

=over 4

Array of full hashes:

    ({ hash => HEX, type => 0 }, { hash => HEX, type => 1 }, { hash => HEX, type => 0 })

=back


=cut

sub get_full_hashes {
	...
}


=head2 update_error()

Save information about a failed database update

	$storage->update_error('time' => time(), wait => 60, errors => 1);


Arguments

=over 4

=item time

Required. Time of the update.

=item wait

Required. Number of seconds to wait before doing the next update.

=item errors

Required. Number of errors.

=back


No return value

=cut

sub update_error {
	...
}

=head2 last_update()

Return information about the last database update

	my $info = $storage->last_update();


No arguments


Return value

=over 4

Hash reference

	{
		time	=> time(),
		errors	=> 0
	}

=back

=cut

sub last_update {
	...
}

=head2 add_full_hashes()

Add full hashes to the local database

	$storage->add_full_hashes(timestamp => time(), full_hashes => [{hash => HEX, list => { }, cache => "300s"}]);


Arguments

=over 4

=item timestamp

Required. Time when the full hash was retrieved.

=item full_hashes

Required. Array of full hashes. Each element is an hash reference in the following format:

	{
		hash		=> HEX,
		list		=> { }',
		cache => "300s"
	}

=back


No return value


=cut

sub add_full_hashes {
	...
}

=head2 full_hash_error()

Save information about failed attempt to retrieve a full hash

	$storage->full_hash_error(timestamp => time(), prefix => HEX);


Arguments

=over 4

=item timestamp

Required. Time when the Google returned an error.

=item prefix

Required. Host prefix.

=back


No return value


=cut

sub full_hash_error {
	...
}

=head2 full_hash_ok()

Save information about a successful attempt to retrieve a full hash

	$storage->full_hash_ok(timestamp => time(), prefix => HEX);


Arguments

=over 4

=item timestamp

Required. Time when the Google returned an error.

=item prefix

Required. Host prefix.

=back


No return value


=cut

sub full_hash_ok {
	...
}

=head2 get_full_hash_error()

Save information about an unsuccessful attempt to retrieve a full hash

	my $info = $storage->get_full_hash_error(prefix => HEX);


Arguments

=over 4

=item prefix

Required. Host prefix.

=back


Return value

=over 4

undef if there was no error

Hash reference in the following format if there was an error:

	{
		timestamp 	=> time(),
		errors		=> 3
	}

=back


=cut

sub get_full_hash_error {
	...
}

=head1 SEE ALSO

See L<Net::Google::SafeBrowsing4> for handling Google Safe Browsing v4.

See L<Net::Google::SafeBrowsing4::Storage::File> for an example of storing and managing the Google Safe Browsing database.

Google Safe Browsing v4 API: L<https://developers.google.com/safe-browsing/v4/>

=head1 AUTHOR

Julien Sobrier, E<lt>julien@sobrier.netE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2016 by Julien Sobrier

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.


=cut

1;


lib/Net/Google/SafeBrowsing4/Storage/File.pm
package Net::Google::SafeBrowsing4::Storage::File;

use strict;
use warnings;

use base qw(Net::Google::SafeBrowsing4::Storage);

use Carp;
use List::Util qw(first);
use Path::Tiny;
use Storable qw(nstore retrieve);
use List::BinarySearch qw(binsearch);


our $VERSION = '0.2';

=head1 NAME

Net::Google::SafeBrowsing4::Storage::File - File storage for the Google Safe Browsing v4 database

=head1 SYNOPSIS

	use Net::Google::SafeBrowsing4::Storage::File;

	my $storage = Net::Google::SafeBrowsing4::Storage::File->new(path => '.');
	$storage->save(...);

=head1 DESCRIPTION

This is the a file-based implementation of L<Net::Google::SafeBrowsing4::Storage> to manage the Google Safe Browsing v4 local database.

=cut


=head1 CONSTRUCTOR

=over 4

=back

=head2 new()

  Create a Net::Google::SafeBrowsing4::Storage::File object

  my $storage	=> Net::Google::SafeBrowsing4::Storage::File->new(path => '.');

Arguments

=over 4

=item path

Optional. Path to store the database files Use current directory by default.

=item keep_all

Optional. Set to 1 to keep old information (such as expiring full hashes) in the database. 0 (delete) by default.

=item sticky

Optional. Set to 1 to if you are going to do multiple lookup. More memory will be used but lookup will be sped up. 0 by default.

=item files

Optional. Hash reference to map file types to file names. Default:

	{
		updates => "updates.gdb4",
		full_hashes => "full_hashes.gsb4"
	}

=item logger

Optional. Log4Perl compatible object reference. By default this option is unset, making Net::Google::SafeBrowsing4::Storage::File silent.

=back

=cut

sub new {
	my ($class, %args) = @_;

	my $self = { # default arguments
		keep_all	=> 0,
		path 		=> '.',
		sticky		=> 0,
		files => {
			updates => "updates.gdb4",
			full_hashes => "full_hashes.gsb4"
		},
		data => { },
		%args,
	};

	bless($self, $class) or croak("Can't bless $class: $!");

	$self->init();

	return $self;
}

=head1 PUBLIC FUNCTIONS

=over 4

See L<Net::Google::SafeBrowsing4::Storage> for a complete list of public functions.

=back

=head2 close()

Cleanup old full hashes, and close the connection to the database.

  $storage->close();

=cut


sub init {
	my ($self, %args) = @_;

	# make sure path exists
	if (! -d $self->{path}) {
		mkdir($self->{path}) or croak("Cannot create directory " . $self->{path} . ": $!\n");
	}

	# file to hold all updates
	my $file = path(join("/", $self->{path}, $self->{files}->{updates}));
	if (! -e $file) {
		my %update = (last_update => 0, next_update => 0, errors => 0);
		if ($self->{sticky}) {
			$self->{data}->{ $self->{files}->{updates} } = { %update };
		}

		nstore(\%update, $file) or croak("Cannot store information into $file: $!\n");
	}
}

sub save {
	my ($self, %args) = @_;
	my $list 		= $args{list} 		|| croak("Missing list information");
	my $override	= $args{override}	|| 0;
	my @hashes		= @{ $args{add} 	|| [] };
	my @remove		= @{ $args{remove} 	|| [] };
	my $state		= $args{'state'}	|| '';

	# save the information somewhere
	my $file = path(join("/", $self->{path}, $self->list_to_file($list)));
	$self->{logger} && $self->{logger}->debug("Save hashes to $file");

	my %data = ('state' => $state, hashes => [@hashes]); # hashes are already stored
	if (-e $file && !$override) {
		my $db = retrieve($file);
		$self->{logger} && $self->{logger}->debug("Load $file (save)");

		$self->{logger} && $self->{logger}->debug("hashes to remove: ", scalar(@remove));
		$self->{logger} && $self->{logger}->debug("hashes to add: ", scalar(@hashes));

		$self->{logger} && $self->{logger}->debug("Number of hashes before removal: ", scalar(@{ $db->{hashes} }));
		foreach my $index (@remove) {
			$self->{logger} && $self->{logger}->debug("Remove index $index");
			$db->{hashes}->[$index] = '';
		}
		$db->{hashes} = [ grep { $_ ne '' } @{ $db->{hashes} } ];
		$self->{logger} && $self->{logger}->debug("Number of hashes after removal: ", scalar(@{ $db->{hashes} }));

		$data{hashes} = [sort { $a cmp $b } (@hashes, @{ $db->{hashes} })];
	}

	nstore(\%data, $file) or croak("Cannot save data to $file: $!\n");
	if ($self->{sticky}) {
		$self->{data}->{ $self->list_to_file($list) } = { %data };
	}

	# return the list of hashes, sorted, from the new storage
	$self->{logger} && $self->{logger}->debug("Number of hashes at end: ", scalar(@{ $data{hashes} }));
	return @{ $data{hashes} };
}



sub reset {
	my ($self, %args) = @_;
	my $list = $args{list} || croak("Missing list information");

	my $file = path(join("/", $self->{path}, $self->list_to_file($list)));
	unlink($file);

	if ($self->{sticky}) {
		$self->{data}->{ $self->list_to_file($list) } = { };
	}
}


sub next_update {
	my ($self, %args) = @_;

	# make sure the file exists
	$self->init();

	my $update = { };
	if ($self->{sticky} && exists($self->{data}->{ $self->{files}->{updates} })) {
		$update = $self->{data}->{ $self->{files}->{updates} };
	}
	else {
		# retrieve information from storage
		my $file = path(join("/", $self->{path}, $self->{files}->{updates}));
		$update = retrieve($file);
		$self->{logger} && $self->{logger}->debug("Load $file (reset)");

		if ($self->{sticky}) {
			$self->{data}->{ $self->{files}->{updates} } = $update;;
		}
	}

	return $update->{next_update} || 0;
}

sub last_update {
	my ($self, %args) = @_;

	# make sure the file exists
	$self->init();

	my $update = { };
	if ($self->{sticky} && exists($self->{data}->{ $self->{files}->{updates} })) {
		$update = $self->{data}->{ $self->{files}->{updates} };
	}
	else {
		# retrieve information from storage
		my $file = path(join("/", $self->{path}, $self->{files}->{updates}));
		$update = retrieve($file);
		$self->{logger} && $self->{logger}->debug("Load $file (last_udpate)");

		if ($self->{sticky}) {
			$self->{data}->{ $self->{files}->{updates} } = $update;
		}
	}

	return { last_update => $update->{last_update} || 0, errors => $update->{errors} || 0 };
}


sub get_state {
	my ($self, %args) = @_;
	my $list = $args{list} || croak("Missing list information\n");

	my $update = { };
	if ($self->{sticky} && exists($self->{data}->{ $self->list_to_file($list) })) {
		$update = $self->{data}->{ $self->list_to_file($list) };
	}
	else {
		my $file = path(join("/", $self->{path}, $self->list_to_file($list)));
		if (! -e $file) {
			return "";
		}
		else {
			$self->{logger} && $self->{logger}->debug("Load $file (get_state)");
			$update = retrieve($file);

			if ($self->{sticky}) {
				$self->{data}->{ $self->list_to_file($list) } = $update;
			}
		}
	}

	return $update->{'state'} || '';
}

sub updated {
	my ($self, %args) = @_;
	my $time = $args{'time'} || time();
	my $next = $args{'next'} || time() + 1800;

	# next update applies to all lists, save it
	# make sure the file exists
	$self->init();

	my $file = path(join("/", $self->{path}, $self->{files}->{updates}));
	my $update = { };
	if ($self->{sticky} && exists($self->{data}->{ $self->{files}->{updates} })) {
		$update = $self->{data}->{ $self->{files}->{updates} };
	}
	else {
		# retrieve information from storage
		$self->{logger} && $self->{logger}->debug("Load $file (updated)");
		$update = retrieve($file);
	}

	$update->{next_update} = $next;
	$update->{last_udpate} = $time;
	$update->{errors} = 0;

	nstore($update, $file) or croak("Cannot save data to $file: $!\n");

	if ($self->{sticky}) {
		$self->{data}->{ $self->{files}->{updates} } = $update;
	}
}


sub update_error {
	my ($self, %args) = @_;
	my $time = $args{'time'} || time();
	my $wait = $args{'wait'} || 1800;
	my $errors = $args{errors} || 0;

	# make sure the file exists
	$self->init();

	my $file = path(join("/", $self->{path}, $self->{files}->{updates}));
	my $update = { };
	if ($self->{sticky} && exists($self->{data}->{ $self->{files}->{updates} })) {
		$update = $self->{data}->{ $self->{files}->{updates} };
	}
	else {
		# retrieve information from storage
		$self->{logger} && $self->{logger}->debug("Load $file (update_error)");
		$update = retrieve($file);
	}

	$update->{next_update} = $time + $wait;
	$update->{last_udpate} = $time;
	$update->{errors}	= $errors;

	nstore($update, $file) or croak("Cannot save data to $file: $!\n");
	if ($self->{sticky}) {
		$self->{data}->{ $self->{files}->{updates} } = $update;
	}
}


sub get_prefixes {
	my ($self, %args) = @_;
	my @lists = @{ $args{lists} || [] };
	my @hashes = @{ $args{hashes} || [] };
	my @data = ();

	$self->{logger} && $self->{logger}->debug("Number of lists: ", scalar(@lists));
	foreach my $list (@lists) {
		my $db = { };
		if ($self->{sticky} && exists($self->{data}->{ $self->list_to_file($list) })) {
			$db = $self->{data}->{ $self->list_to_file($list) };
		}
		else {
			my $file = path(join("/", $self->{path}, $self->list_to_file($list)));
			if (! -e $file) {
				$self->{logger} && $self->{logger}->debug("File $file does not exist");
				next;
			}

			$self->{logger} && $self->{logger}->debug("Load $file (get_prefixes)");
			$db = retrieve($file);

			if ($self->{sticky}) {
				$self->{data}->{ $self->list_to_file($list) } = $db;
			}
		}

		foreach my $hash (@hashes) {
			my $prefix = undef;
			my $index = binsearch {substr($a, 0, length($b)) cmp $b} $hash, @{$db->{hashes}};
			if (defined($index)) {
				$prefix = $db->{hashes}->[$index];
				push(@data, { prefix => $prefix, list => $list, hash => $hash });
			}
		}
	}

	return @data;
}


sub add_full_hashes {
	my ($self, %args) = @_;
	my @hashes = @{ $args{hashes} || [] };
	my $timestamp = $args{timestamp} || time();


	my $file = path(join("/", $self->{path}, $self->{files}->{full_hashes}));
	my $db = { hashes => [] };
	if ($self->{sticky} && exists($self->{data}->{ $self->{files}->{full_hashes} })) {
		$db = $self->{data}->{ $self->{files}->{full_hashes} };
	}
	elsif (-e $file) {
		$db = retrieve($file);
	}

	foreach my $hash (@hashes) {
		my $cache = $hash->{cache};
		$cache =~ s/s//;
		$self->{logger} && $self->{logger}->debug("cache: $cache");

		$hash->{expire} = $cache + $timestamp;
		push(@{ $db->{hashes} }, $hash);
	}

	$self->{logger} && $self->{logger}->debug("Save ", scalar(@{ $db->{hashes} }), " full hashes to $file");
	nstore($db, $file) or croak("Cannot save data to $file: $!\n");

	if ($self->{sticky}) {
		$self->{data}->{ $self->{files}->{full_hashes} } = $db;
	}

	return (@{ $db->{hashes} });
}


sub get_full_hashes {
	my ($self, %args) = @_;
	my @lists = @{ $args{lists} || [] };
	my $hash = $args{hash} || return ();

	my $db = { };
	if ($self->{sticky} && exists($self->{data}->{ $self->{files}->{full_hashes} })) {
		$db = $self->{data}->{ $self->{files}->{full_hashes} };
	}
	else {
		my $file = path(join("/", $self->{path}, $self->{files}->{full_hashes}));
		if (! -e $file) {
			return ();
		}

		$self->{logger} && $self->{logger}->debug("Load $file");
		$db = retrieve($file);
	}

	my @hashes = ();
	$self->{logger} && $self->{logger}->debug("Number of full hashes on file: ", scalar(@{ $db->{hashes} }));
	foreach my $list (@lists) {
		my $result = first {
			$_->{hash} eq $hash &&
			$_->{list}->{threatEntryType} eq $list->{threatEntryType} &&
			$_->{list}->{threatType} eq $list->{threatType} &&
			$_->{list}->{platformType} eq $list->{platformType} &&
			$_->{expire} > time()
		} @{ $db->{hashes} };

		push(@hashes, $result) if (defined($result));
	}

	return @hashes;
}


sub list_to_file {
	my ($self, $list) = @_;

	return join("_", $list->{threatType}, $list->{platformType}, $list->{threatEntryType}) . ".gsb4";
}


sub close {
	my ($self, %args) = @_;

	if ($self->{keep_all} == 0) {
		return;
	}

	my $file = path(join("/", $self->{path}, $self->{files}->{full_hashes}));
	if (! -e $file) {
		return;
	}

	my $db = retrieve($file);

	my @results = grep { $_->{expire} > time() } @{ $db->{hashes} };
	if (scalar(@results) < scalar(@{ $db->{hashes} })) {
		$db->{hashes} = [@results];
		nstore($db, $file) or croak("Cannot save data to $file: $!\n");
	}

	$self->{data} = { };
}

=head1 CHANGELOG

=over 4


=item 0.1

Initial release

=back

=head1 SEE ALSO

See L<Net::Google::SafeBrowsing4> for handling Google Safe Browsing v4.

See L<Net::Google::SafeBrowsing4::Storage> for the list of public functions.

Google Safe Browsing v4 API: L<https://developers.google.com/safe-browsing/v4/>


=head1 AUTHOR

Julien Sobrier, E<lt>julien@sobrier.netE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2016 by Julien Sobrier

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.


=cut

1;